src/main/java/de/zalando/aruha/nakadi/repository/kafka/KafkaPartitionsCalculator.java

```diff
@@ -0,0 +1,150 @@
+package de.zalando.aruha.nakadi.repository.kafka;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.stream.Stream;
+
+public class KafkaPartitionsCalculator {
+    // Contains mapping from message size (in bytes) to list of throughput of kafka with (index + 1) as partition count.
+    private final NavigableMap<Integer, float[]> stats = new TreeMap<>();
+
+    private KafkaPartitionsCalculator(InstanceInfo instanceInfo) {
+        for (SpeedStatistics ss : instanceInfo.getStats()) {
+            stats.put(ss.getMessageSize(), ss.getSpeed());
+        }
+    }
+
+    public int getBestPartitionsCount(int messageSize, float mbsPerSecond) {
+        final Map.Entry<Integer, float[]> floor = stats.floorEntry(messageSize);
+        final Map.Entry<Integer, float[]> ceil = stats.ceilingEntry(messageSize);
+        if (floor == null) {
+            return getBestPartitionsCount(ceil.getValue(), mbsPerSecond);
+        } else if (ceil == null) {
+            return getBestPartitionsCount(floor.getValue(), mbsPerSecond);
+        } else {
+            final int floorResult = getBestPartitionsCount(floor.getValue(), mbsPerSecond);
+            if (Objects.equals(floor.getKey(), ceil.getKey())) {
+                return floorResult;
+            }
+            final int ceilResult = getBestPartitionsCount(ceil.getValue(), mbsPerSecond);
+            return floorResult + (ceilResult - floorResult) * (messageSize - floor.getKey()) / (ceil.getKey() - floor.getKey());
```

Comment:
I think that this rounding is enough. The best we can do is to get better statistics.